engine/src/main/java/org/apache/apex/engine/GroupingManager.java

```diff
@@ -39,7 +40,7 @@
 public class GroupingManager
 {
   private static final GroupingManager groupingManager = new GroupingManager();
-  private Map<String, GroupingRequest> groupingRequests = Maps.newHashMap();
+  private Map<String, GroupingRequest> groupingRequests = Maps.newConcurrentMap();
```

Comment:
Is concurrent necessary? If yes, the pattern used in addOrModifyGroupingRequest is subject to a data loss.